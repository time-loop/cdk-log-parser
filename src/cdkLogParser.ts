import fs from 'fs';

const LOG_REMOVAL_REGEX: [RegExp, string][] = [
  [new RegExp('^\\[Warning at.*\\n', 'gm'), ''],
  [new RegExp('^current credentials could not be used to assume.*\\n', 'gm'), ''],
  [new RegExp('^\\(To get rid of this warning.*\\n', 'gm'), ''],
];

const STACK_REMOVAL_REGEX: [RegExp, string][] = [
  [new RegExp('\\s(└─ \\[~\\] TemplateURL.*[\\s\\S]*?)(?:\\n\\n|\\||\\[ ] ])', ''), ''],
  [new RegExp('.*\\[-] Removed: \\.CDKMetadata\\n\\s', ''), ' '],
  [new RegExp('.*\\[~] .?Metadata:?[\\s\\S]*?└─ \\[\\+].*?(\\n\\s|$)', 'gm'), ' '],
  [new RegExp('^|.*\\n.*├─ \\[~] \\.Metadata:(\\n.*){3}\\n\\s', 'gm'), ''],
  [new RegExp('^(.*)(\\w{64})(\\.zip)\n', 'gm'), ''],
  [
    new RegExp(
      '\\[~] Custom::AWS.*\\n.*Create.*\\n.*getParameter.*\\n.*getParameter.*\\n.*Update.*\\n.*getParameter.*\\n.*getParameter.*',
      'gm',
    ),
    '',
  ],
];

function cleanCDKDiffLog(cdkLog: string) {
  let cleanLog = cdkLog;
  for (const [regex, replacer] of LOG_REMOVAL_REGEX) {
    cleanLog = cleanLog.replace(regex, replacer);
  }
  return cleanLog;
}

function cleanCDKDiffForSingleStack(stackDiff: string) {
  let cleanDiff = stackDiff;
  for (const [regex, replacer] of STACK_REMOVAL_REGEX) {
    cleanDiff = cleanDiff.replace(regex, replacer);
  }

  if (!cleanDiff.match(new RegExp('\\[\\+\\]|\\[\\-\\]', 'gm'))) {
    return 'There were no differences';
  }
  return cleanDiff;
}

function rebuildDiffsIntoCdkLog(stacks: string[]) {
  let cdkLogParts: string[] = [];
  for (const stack of stacks) {
    const stackName = stack.match(new RegExp('(^Stack .*)', 'gm'))?.[0] ?? '';
    const stackDetails = stack.replace(new RegExp('(^Stack .*)\n', 'gm'), '');
    const cleanStackDetails = cleanCDKDiffForSingleStack(stackDetails);
    cdkLogParts.push(stackName);
    cdkLogParts.push(`${cleanStackDetails}\n`);
  }
  return cdkLogParts.join('\n');
}

function separateDiffLogIntoStacks(log: string) {
  const stacks = log.split(new RegExp('(^Stack )', 'gm'));
  if (stacks[0] === '') stacks.shift();
  return stacks.map((_, i) => (i % 2 === 0 ? '' : stacks[i - 1] + stacks[i])).filter((e) => e);
}

function loadCdkDiffLog(filePath: string) {
  return fs.readFileSync(filePath, 'utf8');
}

function saveCdkDiffLog(filePath: string, newStacks: string) {
  fs.writeFileSync(filePath, newStacks);
}

export function run() {
  const cliArgs = process.argv.slice(2);
  if (cliArgs.length !== 1) {
    console.error('Please provide a path to the CDK diff log file');
    return;
  }
  const filePath = cliArgs[0];

  const cdkDiffLog = loadCdkDiffLog(filePath);
  const cleanCdkDiffLog = cleanCDKDiffLog(cdkDiffLog);
  const stacks = separateDiffLogIntoStacks(cleanCdkDiffLog);
  const newStacks = rebuildDiffsIntoCdkLog(stacks);
  saveCdkDiffLog(filePath, newStacks);

  if (newStacks.match(new RegExp('\\[\\+\\]|\\[\\-\\]', 'gm'))) {
    process.exit(1);
  } else {
    process.exit(0);
  }
}

run();
