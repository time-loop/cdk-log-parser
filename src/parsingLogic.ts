import fs from 'fs';

const LOG_REMOVAL_REGEX: [RegExp, string][] = [
  [new RegExp('^\\[Warning at.*\\n', 'gm'), ''],
  [new RegExp('^current credentials could not be used to assume.*\\n', 'gm'), ''],
  [new RegExp('^\\(To get rid of this warning.*\\n', 'gm'), ''],
];

const STACK_REMOVAL_REGEX: [RegExp, string][] = [
  [new RegExp('\\s(└─ \\[~\\] TemplateURL.*[\\s\\S]*?)(?:\\n\\n|\\||\\[ ] ])', ''), ''],
  [new RegExp('.*\\[-] Removed: \\.CDKMetadata\\n\\s', ''), ' '],
  [new RegExp('.*\\[~] .?Metadata:?[\\s\\S]*?└─ \\[\\+].*?(\\n\\s|$)', 'gm'), ' '],
  [new RegExp('^|.*\\n.*├─ \\[~] \\.Metadata:(\\n.*){3}\\n\\s', 'gm'), ''],
  [new RegExp('^(.*)(\\w{64})(\\.zip)\n', 'gm'), ''],
  [
    new RegExp(
      '\\[~] Custom::AWS.*\\n.*Create.*\\n.*getParameter.*\\n.*getParameter.*\\n.*Update.*\\n.*getParameter.*\\n.*getParameter.*',
      'gm',
    ),
    '',
  ],
];

function cleanCdkDiffLog(cdkLog: string) {
  let cleanLog = cdkLog;
  for (const [regex, replacer] of LOG_REMOVAL_REGEX) {
    cleanLog = cleanLog.replace(regex, replacer);
  }
  return cleanLog;
}

function cleanCDKDiffForSingleStack(stackDiff: string) {
  let cleanDiff = stackDiff;
  for (const [regex, replacer] of STACK_REMOVAL_REGEX) {
    cleanDiff = cleanDiff.replace(regex, replacer);
  }

  if (!cleanDiff.match(new RegExp('\\[\\+\\]|\\[\\-\\]', 'gm'))) {
    return 'There were no differences';
  }
  return cleanDiff;
}

function rebuildDiffsIntoCdkLog(stacks: string[]) {
  let cdkLogParts: string[] = [];
  for (const stack of stacks) {
    const stackName = stack.match(new RegExp('(^Stack .*)', 'gm'))?.[0] ?? '';
    const stackDetails = stack.replace(new RegExp('(^Stack .*)\n', 'gm'), '');
    const cleanStackDetails = cleanCDKDiffForSingleStack(stackDetails);
    cdkLogParts.push(stackName);
    cdkLogParts.push(`${cleanStackDetails}\n`);
  }
  return cdkLogParts.join('\n');
}

function separateDiffLogIntoStacks(log: string) {
  const stacks = log.split(new RegExp('(^Stack )', 'gm'));
  if (stacks[0] === '') stacks.shift();
  return stacks.map((_, i) => (i % 2 === 0 ? '' : stacks[i - 1] + stacks[i])).filter((e) => e);
}

export function loadCdkDiffLog(filePath: string) {
  return fs.readFileSync(filePath, 'utf8');
}

export function saveCdkDiffLog(filePath: string, newStacks: string) {
  fs.writeFileSync(filePath, newStacks);
}

export function getParsedStacks(cdkDiffLogContent: string) {
  const cleanedCdkDiffLog = cleanCdkDiffLog(cdkDiffLogContent);
  const stacks = separateDiffLogIntoStacks(cleanedCdkDiffLog);
  return rebuildDiffsIntoCdkLog(stacks);
}
