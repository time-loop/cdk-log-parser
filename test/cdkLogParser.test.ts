import { getParsedStacks, loadCdkDiffLog } from '../src/parsingLogic';

test('no-op', () => {});

describe('getParsedStacks', () => {
  test('should remove all non-sense changes from a serivce instance diff log', () => {
    const cdkDiffLogContent = loadCdkDiffLog('test/data/serviceInstanceOutput.log');
    const parsedStacks = getParsedStacks(cdkDiffLogContent);
    expect(parsedStacks).toMatchSnapshot();
  });

  test('should remove all non-sense changes from a infra diff log', () => {
    const cdkDiffLogContent = loadCdkDiffLog('test/data/infraOutput.log');
    const parsedStacks = getParsedStacks(cdkDiffLogContent);
    expect(parsedStacks).toMatchSnapshot();
  });
});
