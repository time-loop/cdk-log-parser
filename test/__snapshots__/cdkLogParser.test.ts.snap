// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getParsedStacks should ignore content before the first stack diff and NOT clobber the stack names 1`] = `
"Stack InboxProcessingServicePipeline/GlobalProdInboxProcessingServiceDeployHook/DeployHookGlobalProdMonorepoGitRefInboxProcessingServiceProduction (GlobalProdInboxProcessingServiceDeployHook-DeployHookGlobalProdMonorepoGitRefInboxProcessingServiceProduction)
IAM Statement Changes
┌───┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────┬───────────────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────┐
│   │ Resource                                                                                                                                                                                                      │ Effect │ Action                │ Principal                                                                                                                                                                                                                                           │ Condition │
├───┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼───────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────────┤
│ + │ \${GlobalProdInboxProcessingServiceDeployHook/DeployHookGlobalProdMonorepoGitRefInboxProcessingServiceProduction/deploy-hook-global-prod-monorepo-git-ref-inbox-processing-service-production.Arn}             │ Allow  │ lambda:InvokeFunction │ Service:\${GlobalProdInboxProcessingServiceDeployHook/DeployHookGlobalProdMonorepoGitRefInboxProcessingServiceProduction/SsmParameterValue:--ClickUpEcs--Deploy--DeployUser--DeployGlobalProdUserArn:C96584B6-F00A-464E-AD19-53AFF4B05118.Parameter} │           │
└───┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴────────┴───────────────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────┘
IAM Policy Changes
┌───┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────────────────────────────────────────────────────────────────────────────┐
│   │ Resource                                                                                                                                                                                                  │ Managed Policy ARN                                                             │
├───┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────┤
│ + │ \${GlobalProdInboxProcessingServiceDeployHook/DeployHookGlobalProdMonorepoGitRefInboxProcessingServiceProduction/deploy-hook-global-prod-monorepo-git-ref-inbox-processing-service-production/ServiceRole} │ arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole │
└───┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────┘
(NOTE: There may be security-related changes not in this list. See https://github.com/aws/aws-cdk/issues/1299)

Parameters
[+] Parameter GlobalProdInboxProcessingServiceDeployHook/DeployHookGlobalProdMonorepoGitRefInboxProcessingServiceProduction/SsmParameterValue:--ClickUpEcs--Deploy--DeployUser--DeployGlobalProdUserArn:C96584B6-F00A-464E-AD19-53AFF4B05118.Parameter SsmParameterValueClickUpEcsDeployDeployUserDeployGlobalProdUserArnC96584B6F00A464EAD1953AFF4B05118Parameter: {\\"Type\\":\\"AWS::SSM::Parameter::Value<String>\\",\\"Default\\":\\"/ClickUpEcs/Deploy/DeployUser/DeployGlobalProdUserArn\\"}

Resources
[+] AWS::IAM::Role GlobalProdInboxProcessingServiceDeployHook/DeployHookGlobalProdMonorepoGitRefInboxProcessingServiceProduction/deploy-hook-global-prod-monorepo-git-ref-inbox-processing-service-production/ServiceRole deployhookglobalprodmonorepogitrefinboxprocessingserviceproductionServiceRole8BD826CE 

Other Changes
[+] Unknown Rules: {\\"CheckBootstrapVersion\\":{\\"Assertions\\":[{\\"Assert\\":{\\"Fn::Not\\":[{\\"Fn::Contains\\":[[\\"1\\",\\"2\\",\\"3\\",\\"4\\",\\"5\\"],{\\"Ref\\":\\"BootstrapVersion\\"}]}]},\\"AssertDescription\\":\\"CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.\\"}]}}



Stack InboxProcessingServicePipeline/GlobalProdInboxProcessingServiceDeployHook/GlobalProdInboxProcessingServiceDeployHookDeployUserPermissions (GlobalProdInboxProcessingServiceDeployHook-GlobalProdInboxProcessingServiceDeployHookDeployUserPermissions)
There were no differences
"
`;

exports[`getParsedStacks should remove all non-sense changes from a infra diff log 1`] = `
"Stack InboxPipeline/Qa-GitHubActions-Permissions/InboxCdkGithubActions (Qa-GitHubActions-Permissions-InboxCdkGithubActions)
There were no differences

Stack InboxPipeline/Qa-Permissions/InboxPermissions (Qa-Permissions-InboxPermissions)
IAM Statement Changes
┌───┬────────────────────────────────────────────────────────────────────────────┬────────┬─────────────────────────┬────────────────────────────────────────────────────────────────────────────┬───────────┐
│   │ Resource                                                                   │ Effect │ Action                  │ Principal                                                                  │ Condition │
├───┼────────────────────────────────────────────────────────────────────────────┼────────┼─────────────────────────┼────────────────────────────────────────────────────────────────────────────┼───────────┤
│ + │ \${Qa-Permissions/InboxPermissions/inbox-permissions-developer-role-qa.Arn} │ Allow  │ sts:AssumeRole          │ AWS:arn:\${AWS::Partition}:iam::***:user/StrongDM                  │           │
│   │                                                                            │        │                         │ Service:ec2.amazonaws.com                                                  │           │
├───┼────────────────────────────────────────────────────────────────────────────┼────────┼─────────────────────────┼────────────────────────────────────────────────────────────────────────────┼───────────┤
│ + │ *                                                                          │ Allow  │ dynamodb:List*          │ AWS:\${Qa-Permissions/InboxPermissions/inbox-permissions-developer-role-qa} │           │
│   │                                                                            │        │ sqs:List*               │                                                                            │           │
├───┼────────────────────────────────────────────────────────────────────────────┼────────┼─────────────────────────┼────────────────────────────────────────────────────────────────────────────┼───────────┤
│ + │ arn:aws:cloudformation:*:***:stack/*Inbox*                        │ Allow  │ cloudformation:*        │ AWS:\${Qa-Permissions/InboxPermissions/inbox-permissions-developer-role-qa} │           │
├───┼────────────────────────────────────────────────────────────────────────────┼────────┼─────────────────────────┼────────────────────────────────────────────────────────────────────────────┼───────────┤
│ + │ arn:aws:ssm:*:***:parameter/*                                     │ Allow  │ ssm:GetParameter        │ AWS:\${Qa-Permissions/InboxPermissions/inbox-permissions-developer-role-qa} │           │
│   │                                                                            │        │ ssm:GetParameters       │                                                                            │           │
│   │                                                                            │        │ ssm:GetParametersByPath │                                                                            │           │
├───┼────────────────────────────────────────────────────────────────────────────┼────────┼─────────────────────────┼────────────────────────────────────────────────────────────────────────────┼───────────┤
│ + │ arn:aws:ssm:*:***:parameter/Inbox*                                │ Allow  │ ssm:*                   │ AWS:\${Qa-Permissions/InboxPermissions/inbox-permissions-developer-role-qa} │           │
└───┴────────────────────────────────────────────────────────────────────────────┴────────┴─────────────────────────┴────────────────────────────────────────────────────────────────────────────┴───────────┘
(NOTE: There may be security-related changes not in this list. See https://github.com/aws/aws-cdk/issues/1299)

Resources
[+] AWS::IAM::Role Qa-Permissions/InboxPermissions/inbox-permissions-developer-role-qa inboxpermissionsdeveloperroleqaD6B4997D 
[+] AWS::IAM::Policy Qa-Permissions/InboxPermissions/inbox-permissions-developer-role-qa/DefaultPolicy inboxpermissionsdeveloperroleqaDefaultPolicy7153A66A 



Stack InboxPipeline/UsQaIv1/InboxNotificationsDb (UsQaIv1-InboxNotificationsDb)
There were no differences

Stack InboxPipeline/UsQaIv1/InboxNotificationsQueue (UsQaIv1-InboxNotificationsQueue)
There were no differences

Stack InboxPipeline/UsQaQaUsEast21/InboxNotificationsDb (UsQaQaUsEast21-InboxNotificationsDb)
There were no differences

Stack InboxPipeline/UsQaQaUsEast21/InboxNotificationsQueue (UsQaQaUsEast21-InboxNotificationsQueue)
There were no differences

Stack InboxPipeline/UsQaQaUsWest26/InboxNotificationsDb (UsQaQaUsWest26-InboxNotificationsDb)
There were no differences

Stack InboxPipeline/UsQaQaUsWest26/InboxNotificationsQueue (UsQaQaUsWest26-InboxNotificationsQueue)
There were no differences
"
`;

exports[`getParsedStacks should remove all non-sense changes from a service instance diff log 1`] = `
"Stack InboxServicePipeline/Qa-GitHubActions-Permissions/InboxServiceCdkGithubActions (Qa-GitHubActions-Permissions-InboxServiceCdkGithubActions)
There were no differences

Stack InboxServicePipeline/UsQaInboxServiceDeployHook/DeployHookUsQaMonorepoGitRefInboxServiceMaster (UsQaInboxServiceDeployHook-DeployHookUsQaMonorepoGitRefInboxServiceMaster)
There were no differences

Stack InboxServicePipeline/UsQaInboxServiceDeployHook/UsQaInboxServiceDeployHookDeployUserPermissions (UsQaInboxServiceDeployHook-UsQaInboxServiceDeployHookDeployUserPermissions)
There were no differences

Stack InboxServicePipeline/UsQaQaUsEast21/InboxServiceInstance (UsQaQaUsEast21-InboxServiceInstance)
There were no differences

Stack InboxServicePipeline/UsQaQaUsWest26/InboxServiceInstance (UsQaQaUsWest26-InboxServiceInstance)
There were no differences

Stack InboxServicePipeline/UsQaUsWest2/InboxServiceInstance (UsQaUsWest2-InboxServiceInstance)
There were no differences
"
`;
